plugins/analysis-icu/src/main/java/org/elasticsearch/index/mapper/ICUCollationKeywordFieldMapper.java

```diff
@@ -0,0 +1,738 @@
+
+    public static final String CONTENT_TYPE = "icu_collation_keyword";
+
+    public static class Defaults {
+        public static final MappedFieldType FIELD_TYPE = new CollationFieldType();
+
+        static {
+            FIELD_TYPE.setTokenized(false);
+            FIELD_TYPE.setOmitNorms(true);
+            FIELD_TYPE.setIndexOptions(IndexOptions.DOCS);
+            FIELD_TYPE.freeze();
+        }
+
+        public static final String NULL_VALUE = null;
+    }
+
+    public static final class CollationFieldType extends StringFieldType {
+        private final Object collatorLock = new Object();
+        private Collator collator = null;
+
+        public CollationFieldType() {
+            setIndexAnalyzer(Lucene.KEYWORD_ANALYZER);
+            setSearchAnalyzer(Lucene.KEYWORD_ANALYZER);
+        }
+
+        protected CollationFieldType(CollationFieldType ref) {
+            super(ref);
+            this.collator = ref.collator;
+        }
+
+        public CollationFieldType clone() {
+            return new CollationFieldType(this);
+        }
+
+        @Override
+        public boolean equals(Object o) {
+            return super.equals(o) && Objects.equals(collator, ((CollationFieldType) o).collator);
+        }
+
+        @Override
+        public void checkCompatibility(MappedFieldType otherFT, List<String> conflicts, boolean strict) {
+            super.checkCompatibility(otherFT, conflicts, strict);
+            CollationFieldType other = (CollationFieldType) otherFT;
+            if (!Objects.equals(collator, other.collator)) {
+                conflicts.add("mapper [" + name() + "] has different [collator]");
+            }
+        }
+
+        @Override
+        public int hashCode() {
+            return 31 * super.hashCode() + Objects.hashCode(collator);
+        }
+
+        @Override
+        public String typeName() {
+            return CONTENT_TYPE;
+        }
+
+        public Collator collator() {
+            return collator;
+        }
+
+        public void setCollator(Collator collator) {
+            checkIfFrozen();
+            this.collator = collator;
+        }
+
+        @Override
+        public Query nullValueQuery() {
+            if (nullValue() == null) {
+                return null;
+            }
+            return termQuery(nullValue(), null);
+        }
+
+        @Override
+        public IndexFieldData.Builder fielddataBuilder() {
+            failIfNoDocValues();
+            return new DocValuesIndexFieldData.Builder();
+        }
+
+        @Override
+        protected BytesRef indexedValueForSearch(Object value) {
+            if (value == null) {
+                return null;
+            }
+            if (value instanceof BytesRef) {
+                value = ((BytesRef) value).utf8ToString();
+            }
+
+            if (collator != null) {
+                RawCollationKey key;
+                synchronized (collatorLock) {
```

Comment:
I see this in the Collator javadocs: `A frozen Collator is immutable and thread-safe.` Should we freeze the collator and remove the lock?